Zend/zend_language_parser.y

```diff
@@ -422,21 +429,55 @@ is_variadic:
 ;
 
 class_declaration_statement:
-		class_type { $<num>$ = CG(zend_lineno); }
+		class_modifiers T_CLASS { $<num>$ = CG(zend_lineno); }
 		T_STRING extends_from implements_list backup_doc_comment '{' class_statement_list '}'
-			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, $1, $<num>2, $6,
-				  zend_ast_get_str($3), $4, $5, $8); }
-	|	T_INTERFACE { $<num>$ = CG(zend_lineno); }
-		T_STRING interface_extends_list backup_doc_comment '{' class_statement_list '}'
-			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, ZEND_ACC_INTERFACE, $<num>2, $5,
-				  zend_ast_get_str($3), NULL, $4, $7); }
+			{
+				if (!($1 & ZEND_ACC_PPP_MASK)) { $1 |= ZEND_ACC_PUBLIC; }
+				$$ = zend_ast_create_decl(ZEND_AST_CLASS, $1, $<num>3, $7, zend_ast_get_str($4), $5, $6, $9); 
+			}
+	|	T_CLASS { $<num>$ = CG(zend_lineno); }
+		T_STRING extends_from implements_list backup_doc_comment '{' class_statement_list '}'
+			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, ZEND_ACC_PUBLIC, $<num>2, $6, zend_ast_get_str($3), $4, $5, $8); }
+;
+
+class_modifiers:
+		class_modifier 					{ $$ = $1; }
+	|	class_modifiers class_modifier 	{ $$ = zend_add_class_modifier($1, $2); }
+;
+
+class_modifier:
+		T_ABSTRACT 				{ $$ = ZEND_ACC_EXPLICIT_ABSTRACT_CLASS; }
+	|	T_FINAL 				{ $$ = ZEND_ACC_FINAL; }
+	|	visibility_modifier 	{ $$ = $1; }
+;
+
+visibility_modifier:
+		T_PUBLIC 		{ $$ = ZEND_ACC_PUBLIC; }
+	|	T_PRIVATE 		{ $$ = ZEND_ACC_PRIVATE; }
```

Comment:
Could that renamed to class_visibility_modifier ? since the common visibility have the classic PPP modifiers